services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,179 @@
+  }
+
+  private class SystemStoreHealthCheckTask implements Runnable {
+    @Override
+    public void run() {
+      if (!isRunning.get()) {
+        return;
+      }
+
+      Set<String> newUnhealthySystemStoreSet = new HashSet<>();
+      Map<String, Long> systemStoreToHeartbeatTimestampMap = new VeniceConcurrentHashMap<>();
+      for (Store store: storeRepository.getAllStores()) {
+        if (!isRunning.get()) {
+          return;
+        }
+        if (!VeniceSystemStoreUtils.isUserSystemStore(store.getName())) {
+          continue;
+        }
+        // It is also possible that the store is a new store and is being empty pushed.
+        if (store.getCurrentVersion() == 0) {
+          newUnhealthySystemStoreSet.add(store.getName());
+          continue;
+        }
+        VeniceSystemStoreType systemStoreType = VeniceSystemStoreType.getSystemStoreType(store.getName());
+        String userStoreName = systemStoreType.extractRegularStoreName(store.getName());
+        long currentTimestamp = System.currentTimeMillis();
+        if (VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE.equals(systemStoreType)) {
+          pushStatusStoreWriter.writeHeartbeat(userStoreName, currentTimestamp);
+        } else {
+          metaStoreWriter.writeHeartbeat(userStoreName, currentTimestamp);
+        }
+        systemStoreToHeartbeatTimestampMap.put(store.getName(), currentTimestamp);
+      }
+
+      for (Map.Entry<String, Long> entry: systemStoreToHeartbeatTimestampMap.entrySet()) {
+        if (!isRunning.get()) {
+          return;
+        }
+        if (!isSystemStoreIngesting(entry.getKey(), entry.getValue())) {
+          newUnhealthySystemStoreSet.add(entry.getKey());
+        }
+      }
+      LOGGER.info("Collected unhealthy system stores: {}", newUnhealthySystemStoreSet.toString());
+      // Update the unhealthy system store set.
+      unhealthySystemStoreSet.set(newUnhealthySystemStoreSet);
+    }
+  }
+
+  boolean isSystemStoreIngesting(String systemStoreName, long heartbeatTimestamp) {
+    VeniceSystemStoreType systemStoreType = VeniceSystemStoreType.getSystemStoreType(systemStoreName);
+    String userStoreName = systemStoreType.extractRegularStoreName(systemStoreName);
+    try {
+      return RetryUtils.executeWithMaxRetriesAndFixedAttemptDuration(() -> {
+        long retrievedTimestamp;
+        if (systemStoreType == VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE) {
+          retrievedTimestamp = getPushStatusStoreReader()
+              .getHeartbeat(userStoreName, PushStatusStoreUtils.CONTROLLER_HEARTBEAT_INSTANCE_NAME);
+        } else {
+          retrievedTimestamp = getMetaStoreReader().getHeartbeat(userStoreName);
+        }
+        if (retrievedTimestamp < heartbeatTimestamp) {
+          throw new VeniceException("Heartbeat not refreshed.");
+        }
+        return true;
+      }, 3, Duration.ofSeconds(1), Collections.singletonList(VeniceException.class));
```

Comment:
So the healthiness of system store criteria depends on the ordering of the iteration.
For the first store, the heart-beat message needs to be consumed in 3 seconds, otherwise, we would mark the system store as unhealthy, and for the following system stores, the write delay tolerance would be higher.

This may result in some false alarms.

I think we need to define some E2E write latency SLA for system store, and 3 seconds might be too aggressive.
I would like us to make it configurable and maybe we should wait for at least 1 min before querying to avoid false alarms as much as possible.

Reply:
I was thinking that when writing to the list of the system stores I am implicitly giving them enough time to consume the message. 
But if you think this is not enough, then we can wait for 1 minutes. Thread.sleep(1min) should be good enough for the thread. 

But if we sleep for 1 minutes, I don't think we need to have another config to tune the retry 3 seconds? Otherwise, the 1minutes does not make sense.

Reply:
Well, the waiting time is non deterministic.
For example, if the cluster only contains 2 stores, the heart-beat message write won't take more than 2 seconds, that means the gap between sending heart-beat messages and checking the latest heartbeat timestamp will be very short.
That will make it very hard to reason with different types of setup.

I was thinking retry should only be used when client request fails, such as timeout or other errors.